src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsTexture.cpp

```diff
@@ -0,0 +1,348 @@
+                                                 VkImage vulkanImage)
+        : GraphicsTexture(std::move(device), kind, std::move(blockRegion), cpuAccess, width, height, depth),
+          _vulkanImage(vulkanImage),
+          _vulkanImageView([&]() { return CreateVulkanImageView(); }),
+          _vulkanSampler([&]() { return CreateVulkanSampler(); })
+    {
+        VkResult result = vkBindImageMemory(GetAllocator()->GetDevice()->GetVulkanDevice(), vulkanImage,
+                                            GetBlock()->GetVulkanDeviceMemory(), GetBlockRegion().GetOffset());
+
+        if (result != VK_SUCCESS)
+        {
+            throw Exceptions::InitialisationFailureException("Failed to bind VkImage to VkImageMemory correctly.",
+                                                             result);
+        }
+    }
+
+    void* Vulkan::VulkanGraphicsTexture::MapUntyped()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            throw std::runtime_error("Failed to map VkMemory to T*! Reason: " + std::to_string(mapMemoryResult));
+        }
+
+        return pDestination;
+    }
+
+    void* VulkanGraphicsTexture::MapUntyped(size_t rangeOffset, size_t /*rangeLength*/)
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+
+        VkResult result = vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (result != VK_SUCCESS)
+        {
+            throw std::runtime_error("Failed to map VkDeviceMemory! Reason: " + std::to_string(result));
+        }
+
+        return reinterpret_cast<uint8_t*>(pDestination) + rangeOffset;
+    }
+
+    const void* VulkanGraphicsTexture::MapForReadUntyped()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Add real exception.
+            throw std::runtime_error("Failed to map VkMemory to T*! Reason: " + std::to_string(mapMemoryResult));
```

Comment:
Those "real" exceptions got lost on their way here...didn't they?

Reply:
This will probably be another design ticket or part of an existing one that I mentioned previously.